hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -133,14 +141,61 @@ private void initIfNeeded() {
   }
 
   @Override
-  protected List<Pair<String, Option<HoodieRecord<HoodieMetadataPayload>>>> getRecordsByKeys(List<String> keys,
-                                                                                             String partitionName) {
+  public HoodieData<HoodieRecord<HoodieMetadataPayload>> getRecordsByKeyPrefixes(List<String> keyPrefixes,
+                                                                                 String partitionName) {
+    // NOTE: Since we partition records to a particular file-group by full key, we will have
+    //       to scan all file-groups for all key-prefixes as each of these might contain some
+    //       records matching the key-prefix
+    List<FileSlice> partitionFileSlices =
+        HoodieTableMetadataUtil.getPartitionLatestMergedFileSlices(metadataMetaClient, partitionName);
+
+    return engineContext.parallelize(partitionFileSlices)
+        .flatMap(
+            (SerializableFunction<FileSlice, Iterator<Pair<String, Option<HoodieRecord<HoodieMetadataPayload>>>>>) fileSlice -> {
+              // we are moving the readers to executors in this code path. So, reusing readers may not make sense.
+              Pair<HoodieFileReader, HoodieMetadataMergedLogRecordReader> readers =
+                  openReadersIfNeeded(partitionName, fileSlice, false);
```

Comment:
probably we need to fix openReadersIfNeeded for the fullScan/forceFullScan config. with more partitions, may be we want to enable full scan for FILES, but not for other partitions. So, we can't rely on a HoodieMetadataConfig to derive the value for forceFullScan. Each caller might have to set the right value when instantiating the log record reader.

Reply:
Good point. Let's create a ticket to not forget to follow up on it.

Reply:
Chatted offline: only "files" partition could be configured to do full-scan of the logs, while "column_stats", "bloom_filters" will have to go t/h point lookups